features/329-principal-apis/design.md

```diff
@@ -0,0 +1,100 @@
+OAuth. Auth gear should support all identity providers equally well, and
+provide a consistent API for manipulating principals.
+
+
+# Use Cases
+- Get which method is used to log in the current session
+- List all methods to authenticated as a user
+- Add/remove user authentication method
+- Send welcome email to OAuth accounts
+- Populate user metadata according to external profile
+- Create password credentials for users signed up with OAuth
+
+
+# Glossary
+
+**Identity Provider**:
+A method to assert the identity of user. For example, providing a password or
+perform OAuth flow.
+
+**Principal**:
+An identity associated with user. A user can have multiple principals. Since
+the term 'Principal' is defined differently across many frameworks and
+standards, we will instead use a simpler term 'Identity' for public interfaces.
+
+
+# Proposed Design
+
+## Re-authentication for Security
+Many principal manipulation operations are critical for security:
+- Add/remove login ID
+- Link/unlink SSO provider
+- Change password
+
+By default, user are required to re-authenticate before performing these
+operations. This is achieve by checking the issue time of the access token is
+recent.
+
+For change password, we support the common case of providing old password to
+implicitly authenticate the user.
+
+Both the re-auth requirement and time to considered 'recent' can be configured.
+Developer can turn off the re-auth requirement in configuration. However,
+developer should understand the potential security risk.
+
+## Current Principal
+We need to provide a way for developer to get the principal is associated with
+the current session:
+- Which login ID is used to login?
+- Which OAuth provider is used to login?
+
+When logging in using a principal, the ID of the principal (i.e. the current
+principal) would be embedded in the access token along with the user ID.
+Information about the current principal would be returned in user object.
+
+## Principal List
+We need to provide a way for developer to get a list of principals belonged to
+the user:
+- What login IDs the user has?
+- Which OAuth providers had the user linked?
+
+A list of principals belonged to current user would be returned through an API.
+
+# Creating Principal
+We need to provide a way for developer to create new principal for the user:
+- Add new login ID
+- Link OAuth account
+
+Each identity provider should have their specific API to create principal. APIs
+that create principals should have common behaviors:
+- Should respect re-authentication configuration
+
+# Deleting Principal
+We need to provide a way for developer to delete new principal for the user:
+- Remove login ID
+- Unlink OAuth account
+
+Each identity provider should have their specific API to delete principal. APIs
+that delete principals should have common behaviors:
+- Should respect re-authentication configuration
+- Should not allow a verified user to become unverified through the operation
```

Comment:
May you clarify why this behavior is desired?

What if in a case:

1. User A signed up with `a@a.com`
1. User A verified with `a@a.com`
1. User A linked with `b@b.com`
1. User A logged in with `b@b.com` again
1. User A deleted the email address of `a@a.com`

In the case above, user should become unverified? 

Or you intend to enforce the behavior of before `a@a.com` is deleted, `b@b.com` must have been verified? (We need to consider this seriously and what are the implications)

Reply:
It’s just my feeling that it make sense for most app. We can remove this for now and discuss when we specify user verification feature in #336.

Reply:
Sure. Let's postpone it to later / with more real world use cases.